### ISSUE TITLE
m-buffer depends on Emacs25

### ISSUE THREAD

**Issue by josteink** [phillord/m-buffer-el#5]
There are multiple ways to test a major-mode, with [assess.el](https://github.com/phillord/assess) making fontification-tests as simple as they really should be.

To ensure you can run your tests across several Emacs versions, you typically [setup a CI job](https://github.com/josteink/csharp-mode/blob/master/.travis.yml) to provide just that.

And when you've done all that work, just to have all your CI-builds suddenly fail because *something* depends on Emacs 25.0, that's a very sad day.

It turns out assess depends on m-buffer, and [m-buffer now depends on Emacs25.0](https://github.com/phillord/m-buffer-el/commit/9dcfcd69d89dfe464b4994d41ce109c2f218aea9), while in the past it didn't.

Please don't depend on Emacs25.0

It breaks the single most useful use-case for your own test-library :)

**Comment by phillord** [comment]
Oh dear.

I am caught between a rock and a hard place here; my plan is to move
assess into core so that we can improve Emacs tests. So, I have moved
m-buffer to seq from dash. But seq is different between Emacs-25 and
Emacs-24. So, I started the process of transition to Emacs-25.

Still, my intention was not to force Emacs-25 till sometime after it was
actually released. Are you using melpa-stable? The last release version
of assess should be Emacs-24.4 upward, rather than 25, or have I messed
that up?

**Comment by josteink** [comment]
You may have *slightly* misstepped. I currently use MELPA, but if I switch my source to MELPA stable and re-run my test-scripts, I encounter a broken dependency-chain:

````
"/home/jostein/bin/emacs" -Q -batch -L . -l csharp-mode-tests.el -f ert-run-tests-batch-and-exit
Importing package-keyring.gpg...
Importing package-keyring.gpg...done
Contacting host: stable.melpa.org:443
ad-handle-definition: ‘ido-completing-read’ got redefined
Contacting host: stable.melpa.org:443
Contacting host: elpa.gnu.org:80
Package refresh done
Setting ‘package-selected-packages’ temporarily since "emacs -q" would overwrite customizations
Setting ‘package-selected-packages’ temporarily since "emacs -q" would overwrite customizations
Need package ‘m-buffer-0.14’, but only 0.13 is available
make: *** [test] Error 255
````

Checking MELPA stable manually confirms this:

* https://stable.melpa.org/#/assess
* https://stable.melpa.org/#/m-buffer

M-buffer is only version 0.13.

I guess you now have even more things to clean up. Sorry :)

**Comment by josteink** [comment]
Still though. Right now `assess` is not usuable as for CI-builds and all the projects I have which depend on it have broken Emacs 24.x checks.

That's certainly no fun, so any sort of resolution would be appreciated.

**Comment by josteink** [comment]
Checking your own tags in this very project confirms this: `m-buffer` v.0.14 has never been created:

https://github.com/phillord/m-buffer-el/tags

**Comment by phillord** [comment]
Oh dear. You are right, looks like I screwed my tags up. I'll test moving assess back to 0.13 m-buffer, then release a new stable tag for that. This way assess will still use dash based on the old version of m-buffer.

Sorry for hassle.

**Comment by josteink** [comment]
But clearly you had planned a 0.14 release. Can't you check your local tags? Maybe you just forgot a `git push --tags` ?

And can't you check back on your commits, and retro-add the 0.14 tag where you intended to put it in the first place? (Before you started adding Emacs 25-dependant code)?

Anyway. I won't meddle too much. As long as things get sorted out in the end, I'm happy. :)

**Comment by phillord** [comment]
I didn't plan a 0.14 release -- it's just that "0.14-snapshot" versions don't work for some reason I forget. I can't remember why I bumped the dependency version up on assess. It might be because I use assess to test m-buffer.

What a tangled web we weave.

**Comment by phillord** [comment]
Tell a lie, I did plan a 0.14 release. I should have made it immediately before I moved m-buffer to seq.el.

I've added a tag at this point, as you suggest. I'll try and test it later (once melpa-stable has rebuilt!). Fingers crossed it will work.

**Comment by josteink** [comment]
I just tried with the latest 0.14 package, and it's sadly not quite good yet:

````
% emacs --version
GNU Emacs 24.5.1
% git clone https://github.com/josteink/csharp-mode
% cd csharp-mode
% make test
....
Test fontification-of-compiler-directives condition:
    (void-function -zip-pair)
...
3 unexpected results:
   FAILED  fontification-of-compiler-directives
   FAILED  fontification-of-compiler-directives-after-comments
   FAILED  fontification-of-literals-detects-end-of-strings
````

The failing tests are the ones using assess, so there's still something funky going on.

**Comment by phillord** [comment]
Okay, looks like I have the wrong version of dash. -zip-pair is in 2.12.1 not 2.12.0. This is a fix needed in assess. I'll do a new release as soon as I can.

What a pain!

**Comment by phillord** [comment]
Hmmm, been checking. Your test script uses the main ~/.emacs.d/elpa installation for packages right, rather than a clean install of all dependencies. If this is the case, can you delete or update dash.el. This should install 2.12.1 and then all will work.

**Comment by josteink** [comment]
Doesn't seem to work that way. Latest (and only?) version available on MELPA stable is dash 2.13.0, so once again, that's a broken dependency.

But it's a broken dependency I can route around: I've updated my test-runner scripts to manually install dash from MELPA stable, which resolved to 2.13.0, and after that assess installs fine, and now my CI-build is all green again: https://travis-ci.org/josteink/csharp-mode/

I'm not sure what the story with dash is, and if that is something you'd like to investigate further, but for now at least I will consider my problem solved: My assess-based tests runs fine on Emacs 24 :)

Thanks for your persistence in these matters!

**Comment by phillord** [comment]
One of the unfortunate things with package.el is that it can only cope with a single version as far as I know. It's something that needs fixing for any number of reasons. It makes repeatable builds pretty much impossible. So, yes, there will only be a single version of dash.el available.

The last assess failure was because of a missing function that was introduced in 2.12.1 (ironically, I didn't need to use it, but I was worried about a breaking dependency).

Still, I am going to need to think about this. You are right, assess needs to work over several different versions of Emacs. I'm not sure how to achieve this as I want to move things into core; I fear, maintaining a fork is likely to be the only option.
